scrapy_autoextract/providers.py

```diff
@@ -45,17 +46,56 @@ def __init__(
         self.stats = stats
         self.settings = settings
 
+    def __before__(self):
+        providers = self.request.meta.get("_autoextract_providers", [])
+        self.request.meta["_autoextract_providers"] = providers + [self]
+
     async def __call__(self):
         """Make an AutoExtract request and build a Page Input of provided class
         based on API response data.
         """
-        page_type = self.get_page_type()
-        self.stats.inc_value(f"autoextract/{page_type}/total")
+        data = await self._fetch_data()
+        if "error" in data:
+            self.stats.inc_value("autoextract/error/query")
+            raise QueryError(data["query"], data["error"])
+
+        return self.provided_class(data=data)
+
+    @classmethod
+    def register(cls):
+        """Register this provider for its provided class on scrapy-poet
+        registry. This will make it possible to declare provided class as
+        a callback dependency when writing Scrapy spiders.
+        """
+        register(cls, cls.provided_class)
+
+    async def _fetch_data(self):
+        page_type = self.provided_class.item_key
+
+        # When the backend includes support for multiple page types in a single
+        # request, we'll be able to remove this condition and always try to
+        # return the cached data on request's meta.
+        if page_type == "html":
+            # Retrieve cached data on request's meta
+            data = self.request.meta.get("_autoextract_data")
+            if data is not None:
+                return data
```

Comment:
This block is conditioned to the HTML dependency only and there's a reason for it. Currently, AutoExtract backend doesn't support querying multiple page types in a single request. That's why we can't use our cache for Product and Article dependencies, for example. On the other hand, since our "html" page type is merely a placeholder, we can leverage the latest request made since they were made using the "fullHtml" argument set to True.

As soon as it's possible to query multiple page types at once, we should remove the condition here and always seek for cached data on request's meta. We should also refactor the query arguments themselves, but I believe with this approach it will be a very straightforward process.